h/models/group.py

```diff
@@ -190,7 +190,20 @@ def members(self) -> tuple[User, ...]:
         be registered with SQLAlchemy and the changes wouldn't be saved to the
         DB. So this is a read-only property that returns an immutable tuple.
         """
-        return tuple(membership.user for membership in self.memberships)
+        return self.get_members()
+
+    def get_members(self, role: GroupMembershipRoles | None = None) -> tuple[User, ...]:
```

Comment:
This has been changed in the previous PR (https://github.com/hypothesis/h/pull/9047) so that it no longer makes `n + 1` queries.

It's still retrieving _all_ the group's members and memberships from the DB, which is not ideal when this is being called by predicates like `group_has_user_as_owner()` that could do a much smaller query that just checks for a membership with that particular group, user and role.

However, I've tried a few ways to implement that and I'm run into problems, so I'm going to give up:

1. `models.User.get_members()` and use `Session.object_session(self)` to get the SQLAlchemy session and then do specific DB queries. But `object_session()` returns `None` if the ORM object is not attached to any session and this is happening in a bunch of unittests in `predicates_test.py` that don't seem to be easy to fix. There's also [a comment](https://hypothes-is.slack.com/archives/C1MA4E9B9/p1731422994160299) in the predicates tests that seems to suggest that detached instances may happen in the WebSocket app in production too.

2. The security predicate functions themselves can't do DB queries or request services to call (and then the services do DB queries): the predicates don't have access to the request or DB session. The predicates are called by [`identity_permits()`](https://github.com/hypothesis/h/blob/d484af939c1531ed514a5d2b1455ef3b7ea07af1/h/security/permits.py#L9-L42) and `identity_permits()` itself doesn't have access to the request or session, so it'd have to be refactored to take extra arguments and then all the call sites updated, but it's called in many places and some of _them_ don't have the request or session ... It doesn't look particularly easy to do at first glance.

3. [`GroupContext`](https://github.com/hypothesis/h/blob/d484af939c1531ed514a5d2b1455ef3b7ea07af1/h/traversal/group.py#L7-L11) could be updated to have access to `GroupMembersService` and call a new `GroupMembersService.get_members(group, role)` method. The security predicates have access to `GroupContext` so it could provide a forwarding method to `GroupMembersService`. But this seems quite ugly, and when trying to implement it I ran into circular import issues.

Reply:
This method is removed in https://github.com/hypothesis/h/pull/9074